lib/modules/manager/pub/artifacts.spec.ts

```diff
@@ -0,0 +1,265 @@
+import { join } from 'upath';
+import { envMock, mockExecAll } from '../../../../test/exec-util';
+import { env, fs } from '../../../../test/util';
+import { GlobalConfig } from '../../../config/global';
+import type { RepoGlobalConfig } from '../../../config/types';
+import type { UpdateArtifactsConfig } from '../types';
+import * as pub from '.';
+
+jest.mock('../../../util/exec/env');
+jest.mock('../../../util/fs');
+jest.mock('../../../util/git');
+jest.mock('../../../util/http');
+
+const adminConfig: RepoGlobalConfig = {
+  localDir: join('/tmp/github/some/repo'),
+  cacheDir: join('/tmp/cache'),
+  containerbaseDir: join('/tmp/cache/containerbase'),
+};
+
+const config: UpdateArtifactsConfig = {};
+
+describe('modules/manager/pub/artifacts', () => {
+  beforeEach(() => {
+    jest.resetAllMocks();
+
+    env.getChildProcessEnv.mockReturnValue(envMock.basic);
+    GlobalConfig.set(adminConfig);
+  });
+
+  afterEach(() => {
+    GlobalConfig.set(adminConfig);
+  });
+
+  it('returns null if no pubspec.lock found', async () => {
+    const updatedDeps = [{ depName: 'dep1' }];
+    expect(
+      await pub.updateArtifacts({
+        packageFileName: 'pubspec.yaml',
+        updatedDeps,
+        newPackageFileContent: '',
+        config,
+      })
+    ).toBeNull();
+  });
+
+  it('returns null if updatedDeps is empty', async () => {
+    expect(
+      await pub.updateArtifacts({
+        packageFileName: 'pubspec.yaml',
+        updatedDeps: [],
+        newPackageFileContent: '',
+        config,
+      })
+    ).toBeNull();
+  });
+
+  it('returns null if unchanged', async () => {
+    const execSnapshots = mockExecAll();
+    fs.readLocalFile.mockResolvedValueOnce('Current pubspec.lock');
+    fs.readLocalFile.mockResolvedValueOnce('Current pubspec.lock');
+    const updatedDeps = [{ depName: 'dep1' }];
+    expect(
+      await pub.updateArtifacts({
+        packageFileName: 'pubspec.yaml',
+        updatedDeps,
+        newPackageFileContent: '',
+        config,
+      })
+    ).toBeNull();
+    expect(execSnapshots).toMatchObject([
+      {
+        cmd: 'dart pub upgrade',
+        options: {
+          cwd: '/tmp/github/some/repo',
+          encoding: 'utf-8',
+          env: {
+            HOME: '/home/user',
+            HTTPS_PROXY: 'https://example.com',
+            HTTP_PROXY: 'http://example.com',
+            LANG: 'en_US.UTF-8',
+            LC_ALL: 'en_US',
+            NO_PROXY: 'localhost',
+            PATH: '/tmp/path',
+          },
+          maxBuffer: 10485760,
+          timeout: 900000,
```

Comment:
```suggestion
```
remove those values, they are same for all

same for other tests

Reply:
### Docker mode tests

Hopefully I'm on the right direction for the docker mode tests. I tracked down that we probably need to update `allToolConfig` in [containerbase](https://github.com/renovatebot/renovate/blob/main/lib/util/exec/containerbase.ts#L17) and define the tools `Dart` and `Flutter`.

For Flutter, I think the config would be something like this, which seems to be working:

```typescript
flutter: {
  datasource: 'flutter-version',
  depName: 'flutter',
  versioning: npmVersioningId,
},
```

But for Dart, we don't have a data source for it, so I think we would need to add a new data source for Dart. Potentially something similar to this issue: https://github.com/renovatebot/renovate/issues/12820. Would you be able to confirm? If this is the case, I'm also happy to work on adding a data source for Dart.

### Install mode tests

For the install mode tests, I've updated the `GlobalConfig` but it's still not running in install mode (the `exec` snapshot doesn't contain the `install-tool` command). Is there some other config that I need to specify?

<details>
  <summary>Test</summary>
  
  ```ts
  it('supports install mode', async () => {
    GlobalConfig.set({ ...adminConfig, binarySource: 'install' });
    const execSnapshots = mockExecAll();
    fs.getSiblingFileName.mockReturnValueOnce('pubspec.lock');
    fs.readLocalFile.mockResolvedValueOnce('Old pubspec.lock');
    fs.readLocalFile.mockResolvedValueOnce('New pubspec.lock');
    expect(
      await pub.updateArtifacts({
        packageFileName: 'pubspec.yaml',
        updatedDeps: [{ depName: 'dep1' }],
        newPackageFileContent: 'sdk: flutter',
        config,
      })
    ).toEqual([
      {
        file: {
          type: 'addition',
          path: 'pubspec.lock',
          contents: 'New pubspec.lock',
        },
      },
    ]);
    expect(execSnapshots).toMatchObject([
      { cmd: 'install-tool flutter 3.3.9' },
      { cmd: 'flutter pub upgrade' },
    ]);
  });
  ```
</details>